### DISCUSSION TITLE
Could we please improve the uno.extension.navigation docs/gallery? Maybe collect a list of needed content centrally here where more look into than the extensions repo?

### DISCUSSION THREAD

**Discussion by DevTKSS** [unoplatform/uno#19058] [Ideas]
Recognizing much questions about this topic in last days/weeks and also having again and again question marks on my own head from this. The how to section is already big, but I would definitely tell, we need some kind of gallery area for this, e.g. not only the actual "look to the left here and see the NavigationView, for more information/samples see the code" (raw transferred) more like the <> areas of other controls. It's really hard to understand this in detail.
If others also agree we could do some kind of list, which information we thing to be missing, maybe see they are existing but some links could be added in this cases to provide a better overview where to find what...

For example I am thinking about:
- Provide a full list of possible Map (ViewMap, DataViewMap...) including sample for c# markup beneath xaml
- what can we actually configure on navigation? Like the debugging output? Other things? Did not see possibility in the app.xaml.cs commented lines
- full list of RouteMap (is it the only method useable at the route registry view section below the ViewMaps? Looked up the definition but that is really telling nothing about what methods are supported and copilot in vs is suggesting totally wrong(not provided) methods, while intellisense is not suggesting anything.
- information about navigationView also! Maybe sounds doubled up, but the docs are telling different things which are confusing and i only did find xaml definition using uen: for using it. E.g. how would this look like when we would like to navigate from mainModel (task bound to command) using mvux (no docs found on this with uno navigator, am I blind?) in the c# code to the nested page? I would assume from the docs, it could be "navigateRouteAsync(this,"MainPage/NestedNextPage") but trying this is not compiling.
- how to work with DataContext with navigation? Doing just "new MainViewModel();" is known to throw NullReferenceException issues which is disabling to use x:Bind to commands on the model for example

As said, that would just be some examples I would maybe add to such to-do list if there would be one :) seen that in the actual repo on this there are some questions. We could also kind of pin such list or link to list (clear described what it is about to enable a better linking for new ones here also) here on the repo to not let it go under?
Yes I know, much things in one idea, but in my opinion it should be hold together for better overview and following up...

Looking for your opinions!

**Comment by kazo0** [comment]
Hi @DevTKSS, thanks for the feedback!

I'll try and summarize the current Navigation documentation/guidance and I'd like to get a better understanding of what you think my be missing.

For step-by-step guides on using Navigation for different scenarios we have the [How-To docs for Navigation](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/NavigationOverview.html)

<img width="332" alt="image" src="https://github.com/user-attachments/assets/282afedc-afcb-4bc7-9796-752d9e8e3dbe" />

As a supplement to that, we have a [Navigation reference sample app](https://github.com/unoplatform/Uno.Samples/tree/master/UI/Navigation) where we are implementing as many of the Navigation scenarios as possible.

For you bullet list of examples:

> Provide a full list of possible Map (ViewMap, DataViewMap...) including sample for c# markup beneath xaml

and

> full list of RouteMap (is it the only method useable at the route registry view section below the ViewMaps...

- For these two points, we have the [Route Guidance here
](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-DefineRoutes.html), are you requesting further explanation or just that we should also provide the syntax to use Routes with C# Markup?

- Would it be more beneficial to provide a breakdown of all methods and properties that exist on the Route/Map types? Similar to how we are documenting types in the Toolkit documentation ([see here](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/DrawerControl.html#constructors) for example)

> what can we actually configure on navigation? ...

- This is interesting, are you asking for something in the documentation that explains the different signatures available for the `UseNavigation()` extension methods [found here](https://github.com/unoplatform/uno.extensions/blob/44e66c13fcbafc6817863ccb96a875eaef8bd526/src/Uno.Extensions.Navigation.UI/HostBuilderExtensions.cs#L26)? 

> information about navigationView [...] how would this look like when we would like to navigate from mainModel (task bound to command) using mvux

- For this we have current documentation on [Navigating via ViewModel](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-NavigateInCode.html#4-view-model-navigation) 
- As long as your routes are set up properly for nested pages, you should be able to navigate via ViewModel directly to your "NestedNextPage" route. Similar to the Route definition in the [Navigation to Page Elements section](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/Advanced/HowTo-UseNavigationView.html#navigating-to-page-elements), you "NestedNextPage" would be the equivalent to the "Products" route defined in the docs. So simply navigating to the "NestedNextPage" should properly navigate forward from the MainPage

> how to work with DataContext with navigation? Doing just "new MainViewModel();" is known to throw NullReferenceException issues which is disabling to use x:Bind to commands on the model for example

- The creation of the DataContext should be handled for you by the Navigation extensions. The [ViewMap section](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-DefineRoutes.html#viewmap) of the docs is where we explain that the given route/view is registered with a specific ViewModel type and upon navigating to a given route, the Navigation extensions will automatically resolve the proper ViewModel and set it as the DataContext.

Let me know if you need more information on these points, and if you had further ideas on how to improve the Navigation documentation!

**Reply by DevTKSS** [reply]
> For step-by-step guides on using Navigation for different scenarios we have the [How-To docs for Navigation](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/NavigationOverview.html)

I am already following this guides, but because I did not find what I would maybe need for fully understand and use I added this idea for docs improvement, because seeing the questions also asked from others beside me, that might be interesting also for others to know.
 
> As a supplement to that, we have a [Navigation reference sample app](https://github.com/unoplatform/Uno.Samples/tree/master/UI/Navigation) where we are implementing as many of the Navigation scenarios as possible.

Could we do a working (cloned the navigation sample app already and it had problematic (outdated) dependencys...
Maybe it would be possible to add the navigation app/content as samples in the uno gallery app? at the moment we only having is the NavigationView beside the NavigationBar which is also not providing any code sample in the uno-gallery app:
![image](https://github.com/user-attachments/assets/62ccd4bc-d176-439a-b945-92248c0a4c4b)
And before you tell, yes, I also know the navigationView & NavigationBar content in the docs ;) this one here is only the idea of adding samples+code to expand to the gallery for navigation. the ui elements are nice to see, but with code for all uno derrived design patterns would be nice also to include there :)

> For you bullet list of examples:
> 
> > Provide a full list of possible Map (ViewMap, DataViewMap...) including sample for c# markup beneath xaml
> 
> and
> 
> > full list of RouteMap (is it the only method useable at the route registry view section below the ViewMaps...
> 
> * For these two points, we have the [Route Guidance here
>   ](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-DefineRoutes.html), are you requesting further explanation or just that we should also provide the syntax to use Routes with C# Markup?

since I seem to not be able to use uno for maybe needed .net framework at work, I am trying to at least learn xaml from it while doing my own project for autmating some tasks. So since WPF (seems like I will have to use it for my job in future) is xaml driven like uno, no c# is not my first need, but I would not tell, others maybe need this in c# markup also, maybe a side-by side tab driven thing like in the Dialogs page? Had done my first project in c# markup and runned into thread issue (linked below) so given up on c#, DropdownButton and ContentDialog via Navigator just then...

To the Routes, like on the HostBuilder Extensions, I would love to have a full table of possible Methods we can use, could have expanders added for samples. that maybe would be also a really nice thing if possible to have in the gallery mentioned above?

> * Would it be more beneficial to provide a breakdown of all methods and properties that exist on the Route/Map types? Similar to how we are documenting types in the Toolkit documentation ([see here](https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/DrawerControl.html#constructors) for example)
> 
> > what can we actually configure on navigation? ...
> 
> * This is interesting, are you asking for something in the documentation that explains the different signatures available for the `UseNavigation()` extension methods [found here](https://github.com/unoplatform/uno.extensions/blob/44e66c13fcbafc6817863ccb96a875eaef8bd526/src/Uno.Extensions.Navigation.UI/HostBuilderExtensions.cs#L26)?

Yes, think thats what was showing on the app.xaml.cs, but actually I have not found any documentation for what we all can do/configure there. I have seen on others debug output, they get shown, when the navigator had done what, what maybe would be a bit helpfull for understanding, how I can estimate the DataContext to be set or what is going on there what is causing my exception mentioned below in seperate thread in this post. I would like to understand what is going on to maybe see my own issues made from not understand something correct. Normally if it comes to microsoft things I would lookup their documentation and lookup the class/Interface or whatever I need then. The Reference Uno Docs have is... could be improved :) 
But do you thing its possible to add more detail into docs that would provide something similar like the reference microsoft is doing, just in this case the object would be the uno extensions and also the appbuilder. Its there, its preconfigured and the wizard is nice, no doubt, but if we think about adding/changing parts there, for me it feels limited to the commented things you see in your prebuild app.
 
> > information about navigationView [...] how would this look like when we would like to navigate from mainModel (task bound to command) using mvux
> 
> * For this we have current documentation on [Navigating via ViewModel](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-NavigateInCode.html#4-view-model-navigation)

in my understandment, ViewModel is not equal to MainModel, that would be MainViewModel. would tell this to be correct reading this on "How-To: Define Routes"
![image](https://github.com/user-attachments/assets/590fe89a-0eff-4000-b866-e4d42e7e1efe)
so when I normally use MVUX, the MainViewModel would be generated, so I would need the "How-To: Navigate via Model" to be added with full list of possibilitys or also not able methods? its failing when NavigateViewModelAsync(this, "Nested", qualifyers: Qualifyers.Nested) possible I should use NavigateRouteAsync, but thats the start of I do simply not know how to navigate right there.

> * As long as your routes are set up properly for nested pages, you should be able to navigate via ViewModel directly to your "NestedNextPage" route. Similar to the Route definition in the [Navigation to Page Elements section](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/Advanced/HowTo-UseNavigationView.html#navigating-to-page-elements), you "NestedNextPage" would be the equivalent to the "Products" route defined in the docs. So simply navigating to the "NestedNextPage" should properly navigate forward from the MainPage
>

Thats the best I got (but this is using mvvm because mvux is most of the time failing, tryed again because of the hotdesign guide for my current application.
https://github.com/DevTKSS/UnoApp1.git
I wanted to do a small repro, but while trying to move from c: it seems like I broken dotnet and compiler, which is not longer finding uno things in my app, but uno-check is telling all fine XD so sorry for not beeing able to provide that to maybe see the points we might need improvement. I could only try switching my hotDesign try project to public, but not more...
For example, in the project I linked, on MainPage I added the NavigationView and added the names along the guide for this.

Not clear for me, especially on the xaml part of navigation:
we have Region.Name, Region.Attached and Region.Navigator which are regulary used in this guides, but when it comes to the uen:Navigator.Data or Request, i only found [this](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-SelectValue.html) but I do not have any idea how they exchanged/bound the request. I do see the qualifyer "-" which indicates we want to go back in the navigationStack, but I see no command or similar at all, that would bind the actual navigator Task to run?

- also [here](https://platform.uno/docs/articles/external/uno.extensions/doc/Reference/Navigation/Qualifiers.html), when we are at the qualifyers, could be added examples, because I did try that, see my initial bullet 4, which would replicate the "Example Usage" from there, but did not at all work.

> > how to work with DataContext with navigation? Doing just "new MainViewModel();" is known to throw NullReferenceException issues which is disabling to use x:Bind to commands on the model for example
> 
> * The creation of the DataContext should be handled for you by the Navigation extensions. The [ViewMap section](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-DefineRoutes.html#viewmap) of the docs is where we explain that the given route/view is registered with a specific ViewModel type and upon navigating to a given route, the Navigation extensions will automatically resolve the proper ViewModel and set it as the DataContext.
> 
See my added thread on this post for DataContext issues on Navigator/Mvux/Binding

would like to add to the list, checking out the repo you linked from the Hostbuilder:
![image](https://github.com/user-attachments/assets/7bcdf852-13b7-4e50-b132-866434ce5fc2)

- seeing this quite cool additional Navigators, did not see any docs about them, was told I have to use NavigateMessageDialogAsync, while at the Dialogs Page is told, to avoid the native MessageDialog thats not really good sign for beginners, to first tell, not to use something and then name the Method to use exactly like this XD thats a bit ironic in my opinion. Is it actually possible, and how would that be done, to use these other Navigators? Maybe I would like to have one to do the Dialogs, dont know? Would appreciate it to only have to set the xaml root for this only once 🤣 seeing my problems with this thing in my first Uno app that was a realistic idea, just if I would have had any idea how to do 😄

**Comment by DevTKSS** [comment]
@kazo0 Existing Issue for DataContext as Example:
https://github.com/unoplatform/uno.extensions/issues/2629
related to my approach (but my command is not even using x:Bind twoWay!) #12073
seems like its on the Bind a general problem, which makes ist even harder to use or only with Binding.
My Idea has been, to manually generate the Bind eventdriven, that would normally be done via xaml but for this I do not know how to do this in a xaml markup application. Only Binding things documentated are 90% using Binding and the ones using x:Bind (not beeing c# ui markup) are the things binding to controls on same pages but for this I am failing on how to set this without codebehind (understood with mvux this should tryed to eliminate?) Click on button => NavigationView.IsPaneOpen just in xaml. code not worked.
so coming from the exception, I thought, when we seem to have a `this.Navigator()?.Get<SomeType>();` along documentation, I could try to let the type be MainViewModel, because Navigator dont lets me set the constructor to take the interfaces MainModel needs and makes MainViewModel needs them too, it is requiring MainPage to be Parameterless Constructor. so, when I am reading the docs, hearing your answer:

> The creation of the DataContext should be handled for you by the Navigation extensions. The [ViewMap section](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Navigation/HowTo-DefineRoutes.html#viewmap) of the docs is where we explain that the given route/view is registered with a specific ViewModel type and upon navigating to a given route, the Navigation extensions will automatically resolve the proper ViewModel and set it as the DataContext.

I would please need a working way to do this in the docs with using the navigator 👍
If there is no way, that statement of "Navigator is setting DataContext" in the Docs should be clarifyed to be not true when it comes to x:Bind in my opinion.
Reading this Doc Page here:
https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/DependencyInjection/HowTo-DependencyInjectionSetup.html
on the 5. Part right on the bottom of the page, they are, as you told, telling the thing with DataContext beeing set from Navigator. Also should be made clear for this case. Its Okay, that on a openSource Repo like uno, not everything is working all the time, but could we please fit the docs for this to be clearer on this bullet point?
here on Troubleshoot is also not listed something like this: https://platform.uno/docs/articles/uno-builds-troubleshooting.html

**Reply by DevTKSS** [reply]
Would make sense to collect all the issues also since they keep popping up with very similar source. Mostly its the bind/binding to the property MainViewModel or other viewmodels. Found about 10 in last 2 days

**Comment by DevTKSS** [comment]
@kazo0 

https://platform.uno/docs/articles/external/uno.extensions/doc/Reference/Navigation/Design.html
maybe we could get the list of Navigation Controls linked to the specific guide?
I did find the docs you linked here already and studied them, but its hard sometimes to find the areas searched for, especially if someone is new to uno and all the features. they are kind of spread everywhere.

and as beginner, where I find this: https://platform.uno/docs/articles/features/dialogs.html?tabs=mvux%2Cxaml
it was not clear at all, how to do the c# mvux approach (now switched to xaml markup) with an app that is using Navigator. Until today my issue resulting from this try has not stepped any future as far as I know. it throwed an issue that a non ui Element is used on a non ui thread but I would have no Idea what I had done wrong. https://github.com/unoplatform/uno.extensions/issues/2618
It is realistic, that its just a beginner fault, so I would maybe also want to add the content to set inside of the task which is doing this contentDialog in a MainModel driven by the Navigator to be added if possible
